extractor/text.go

```diff
@@ -848,38 +912,255 @@ func (pt PageText) length() int {
+func rectUnion(b1, b2 model.PdfRectangle) model.PdfRectangle {
+	return model.PdfRectangle{
+		Llx: math.Min(b1.Llx, b2.Llx),
+		Lly: math.Min(b1.Lly, b2.Lly),
+		Urx: math.Max(b1.Urx, b2.Urx),
+		Ury: math.Max(b1.Ury, b2.Ury),
+	}
+}
+
+// TextMark represents extracted text on a page with information regarding both textual content,
+// formatting (font and size) and positioning.
+// It is the smallest unit of text on a PDF page, typically a single character.
+//
+// getBBox() in test_text.go shows how to compute bounding boxes of substrings of extracted text.
+// The following code extracts the text on PDF page `page` into `text` then finds the bounding box
+// `bbox` of substring `term` in `text`. (indexRunes() works like strings.Index except that it
+// returns number of runes rather than numberof bytes)
+//
+//     ex, _ := New(page)
+//     // handle errors
+//     pageText, _, _, err := ex.ExtractPageText()
+//     // handle errors
+//     text := pageText.Text()
+//     textMarks := pageText.Marks()
+//
+//     start := indexRunes(text, term)
+//     end := start + len([]rune(term))
+//     spanMarks, err := textMarks.RangeOffset(start, end)
+//     // handle errors
+//     bbox, ok := spanMarks.BBox()
+//     // handle errrors
+type TextMark struct {
+	// Text is the extracted text. It has been decoded to Unicode via ToUnicode().
+	Text string
+	// Original is the text in the PDF. It has not been decoded like `Text`.
+	Original string
+	// BBox is the bounding box of the text.
+	BBox model.PdfRectangle
+	// Font is the font the text was drawn with.
+	Font *model.PdfFont
+	// FontSize is the font size the text was drawn with.
+	FontSize float64
+	// Offset is the offset of the start of TextMark.Text in the extracted text. If you do this
+	//   text, textMarks := pageText.Text(), pageText.Marks()
+	//   marks := textMarks.Elements()
+	// then marks[i].Offset is the offset of marks[i].Text in text.
+	Offset int
+	// Meta is set true for spaces and line breaks that we insert in the extracted text. We insert
+	// spaces (line breaks) when we see characters that are over a threshold horizontal (vertical)
+	//  distance  apart. See wordJoiner (lineJoiner) in PageText.computeViews().
+	Meta bool
+}
+
+// String returns a string describing `tm`.
+func (tm TextMark) String() string {
+	b := tm.BBox
+	var font string
+	if tm.Font != nil {
+		font = tm.Font.String()
+		if len(font) > 50 {
+			font = font[:50] + "..."
+		}
+	}
+	var meta string
+	if tm.Meta {
+		meta = " *M*"
+	}
+	return fmt.Sprintf("{TextMark: %d %q=%02x (%5.1f, %5.1f) (%5.1f, %5.1f) %s%s}",
+		tm.Offset, tm.Text, []rune(tm.Text), b.Llx, b.Lly, b.Urx, b.Ury, font, meta)
 }
 
-// ToText returns the contents of `pt` as a single string.
-func (pt PageText) ToText() string {
+// computeViews processes the page TextMarks sorting by position and populates `pt.viewText` and
+// `pt.viewMarks` which represent the text and marks in the order which it is read on the page.
```

Comment:
This is a breaking change (changing already exported signature), need to leave the ToText() string

Reply:
Can leave ToText() which returns Text() and a note that ToText will be deprecated in next major version.